### DISCUSSION TITLE
d8ahazard, please add info and examples about [filewords]

### DISCUSSION THREAD

**Discussion by manniacmind** [d8ahazard/sd_dreambooth_extension#844] [Ideas]
I tried to find out what [filewords] are for and how exactly they are used. i looked at different threads in discussions and different people claim different things. It seems to be impossible to find out who is right. Also my own tests with the feature did not yield good results. It seems like the training results are way worse when using [filewords], so probably I did something wrong or misunderstood.

So, please d8ahazard, can you add more information how we are supposed to use [filewords] and also give specific examples?

- what exact text should go into the instance token and the class token fields, when using [filewords]?
- what exact text should go into the class images txt-files, when using [filewords]?
- what does the script actually do with [filewords]? how does it change text from the txt-file, or the instance token or the class token fields? Does it remove words? Does it add words? Does it replace them? What does shuffling do?
- how does using [filewords] benefit us? does it have disadvantages? does it speed up or slow down the training? is the training result getting better or worse? is inference getting more flexible when using [filewords]? What exactly should the txt-description include and what should it leave out? I read that it should only include things about the subject that should remain flexible. for instance when the character has always red hair, then red hair should not be part of the description. But if the hair STYLE varies, then it should include "long hair" and "short hair" etc. Is this right?

**Comment by Zuxier** [comment]
did you read the wiki? is there for that you know

**Reply by FugueSegue** [reply]
I have read the Wiki several times. There are no examples of the proper use of [filewords]. It only states that it can be done. Not how it's done. There needs to be examples of all the variations of how it's used. There is some information about that in the Wiki. But IMHO there is not enough.

I found out only yesterday that you do not include the instance token within the text of the caption files. Only the class token should be in each caption file. Or is this correct? I don't know.
[Video where I learned this.
](https://youtu.be/KwxNcGhHuLY?t=378)
One question I have is about commas within caption files. Is it a bad idea? I don't know.

**Reply by Zuxier** [reply]
> I have read the Wiki several times. There are no examples of the proper use of [filewords]. It only states that it can be done. Not how it's done. There needs to be examples of all the variations of how it's used. There is some information about that in the Wiki. But IMHO there is not enough.
> 
> I found out only yesterday that you do not include the instance token within the text of the caption files. Only the class token should be in each caption file. Or is this correct? I don't know. [Video where I learned this. ](https://youtu.be/KwxNcGhHuLY?t=378) One question I have is about commas within caption files. Is it a bad idea? I don't know.

i will go the extra mile, just to prove in fact that the wiki explains all of the questions above
> 
>what exact text should go into the instance token and the class token fields, when using [filewords]?`

![image](https://user-images.githubusercontent.com/120954436/214637057-5b24681f-7a99-4465-8aab-d6ecdb683f5e.png)

> 
>what exact text should go into the class images txt-files, when using [filewords]?

![image](https://user-images.githubusercontent.com/120954436/214637350-27ecb30e-92e9-4577-ae6d-c7fe93371c02.png)

> 
>what does the script actually do with [filewords]? how does it change text from the txt-file, or the instance token or the class token fields? Does it remove words? Does it add words? Does it replace them? What does shuffling do?

![image](https://user-images.githubusercontent.com/120954436/214637886-7319d1f2-92ba-4d79-820a-e7604765afe5.png)
shuffling is right now missing from the wiki, but what it does is mixing the tags every epoch to make prompting with the final model more flexible

>
>how does using [filewords] benefit us? does it have disadvantages? does it speed up or slow down the training? is the training result getting better or worse? is inference getting more flexible when using [filewords]? What exactly should the txt-description include and what should it leave out? I read that it should only include things about the subject that should remain flexible. for instance when the character has always red hair, then red hair should not be part of the description. But if the hair STYLE varies, then it should include "long hair" and "short hair" etc. Is this right?

the response to this question is a mix of all of the above responses, by using filewords you are basically training multiple words at the same time, so hair get closer to the hair that are being trained etc etc.

I wrote you on discord as well btw, in discord there is stuff for days.

**Reply by manniacmind** [reply]
From your answer I understand that
- instance token should be "ohwx"
- class token should be "person"
- instance prompt should be "[filewords]"
- class prompt should be "[filewords]"
- txt-files should contain something like "man standing next to chair, red chair"
Is this correct?

Because I was under the impression that the instance prompt should be "a photo of ohwx person, [filewords]" and the class prompt should be "a photo of a person, [filewords]".
Is this wrong?

Also, can/should I use "ohwx" and "person" in the txt-files?

You didn't write me on discord. Maybe you confused me? I'm not GunnarQuist, but the original poster of this thread.

**Reply by Zuxier** [reply]
> Because I was under the impression that the instance prompt should be "a photo of ohwx person, [filewords]" and the class prompt should be "a photo of a person, [filewords]". Is this wrong?

nowhere in the wiki, it's explained be used like that.
Instance token are prepended to instance prompts during instance image training.
Class tokens are prepended to class prompts during class image generation
So ohwx/person doesn't need to be into the filewords.
instance and class prompt can both be [filewords].
let's say filewords for image1 contains: `white hair, happy, blue shirt`.
image1 will be trained with `ohwx, white hair, happy, blue shirt`.
classimage1 will be generated with `person, white hair, happy, blue shirt`.

**Reply by manniacmind** [reply]
thank you. But is this example, that I gave, correct? yes or no?
![image](https://user-images.githubusercontent.com/119996840/214723295-70d48f09-96a5-4b93-88c5-77e923b1a4e8.png)

**Reply by manniacmind** [reply]
> nowhere in the wiki, it's explained be used like that.

That's why I asked for examples. It's very difficult to learn these things without examples.

**Comment by NeuralBlankes** [comment]
the text files should *probably* be like that, yes.  The training side would couple the instance token with your filewords to make: "ohwx man standing next to chair, red chair"
That said, the training I'm running now apparently isn't actually pulling the filewords.  When it generates a sample, it generates a text file with each sample of the prompt used, and I'm getting "[filewords]" as the prompt, instead of the actual text held within filewords file.  This is also reflected in the sample images produced.
